src/main/scala/com/fulcrumgenomics/bam/FindSwitchbackReads.scala

```diff
@@ -0,0 +1,398 @@
+ *
+ * Copyright (c) 2018 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import java.lang.Math.{abs, max, min}
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.bam.api.{SamOrder, SamRecord, SamSource, SamWriter}
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.commons.io.PathUtil
+import com.fulcrumgenomics.commons.util.{LazyLogging, NumericCounter}
+import com.fulcrumgenomics.fasta.ReferenceSequenceIterator
+import com.fulcrumgenomics.sopt.{arg, clp}
+import com.fulcrumgenomics.util.Metric.{Count, Proportion}
+import com.fulcrumgenomics.util._
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools.SamPairUtil.PairOrientation
+import htsjdk.samtools.reference.ReferenceSequence
+import htsjdk.samtools.util.SequenceUtil
+import htsjdk.samtools.{CigarOperator, SAMFileHeader, SAMReadGroupRecord}
+
+
+object FindSwitchbackReads {
+  /** Resource path for R script for plotting. */
+  val PlottingScript = "com/fulcrumgenomics/bam/FindSwitchbackReads.R"
+
+  /** The name of the BAM extended attribute tag used to store switchback information. */
+  val SwitchTag = "sb"
+
+  /** Parent trait for the different kinds of hits we can find. */
+  sealed trait SwitchHit { def code: Char }
+
+  /**
+    * Information about a switchback found by looking at soft-clipped complementary sequence within
+    * a single read.
+    *
+    * @param offset the offset between the reference position at the last base read on the original strand and
+    *               the first base read on the opposite strand.
+    * @param length the length of the post-switch sequence within the read
+    * @param read optionally the read in which the hit was found (generally filled in later)
+    */
+  case class ReadBasedHit(var offset: Int, length: Int, var read: Option[SamRecord] = None) extends SwitchHit {
```

Comment:
I'd prefer `offset` and `read` to be `val`s.  Is it possible to update the method that creates a "hit" to make it so?

Reply:
I don't think it'd be _that_ complicated, since your just adding some parameters to a function or two, but since I am just a commentator, I'll let it go.